linera-execution/src/system.rs

```diff
@@ -711,42 +737,48 @@ where
                     Recipient::Burn => (),
                 }
             }
-            // These messages are executed immediately when cross-chain requests are received.
-            OpenChain(_) => {}
             // This message is only a placeholder: Its ID is part of the application ID.
             ApplicationCreated => {}
         }
         Ok(outcome)
     }
 
     /// Initializes the system application state on a newly opened chain.
-    pub fn initialize_chain(
-        &mut self,
-        message_id: MessageId,
-        timestamp: Timestamp,
-        config: OpenChainConfig,
-    ) {
+    pub async fn initialize_chain(&mut self, chain_id: ChainId) -> Result<(), ExecutionError> {
         // Guaranteed under BFT assumptions.
         assert!(self.description.get().is_none());
         assert!(!self.ownership.get().is_active());
         assert!(self.committees.get().is_empty());
-        let OpenChainConfig {
+        let description_blob = self
+            .read_blob_content(BlobId::new(chain_id.0, BlobType::ChainDescription))
+            .await?;
+        let description: ChainDescription = bcs::from_bytes(description_blob.bytes())?;
+        let InitialChainConfig {
             ownership,
             admin_id,
             epoch,
             committees,
             balance,
             application_permissions,
-        } = config;
-        let description = ChainDescription::Child(message_id);
+        } = description.config().clone();
+        self.timestamp.set(description.timestamp());
         self.description.set(Some(description));
         self.epoch.set(Some(epoch));
+        let committees = committees
+            .into_iter()
+            .map(|(epoch, serialized_committee)| {
+                let committee = bcs::from_bytes(&serialized_committee)
+                    .expect("Deserializing a committee shouldn't fail");
+                (epoch, committee)
+            })
+            .collect();
         self.committees.set(committees);
-        self.admin_id.set(Some(admin_id));
+        // If `admin_id` is `None`, this chain is its own admin chain.
```

Comment:
Right, nice! (Ignore my other comment about making this non-optional; doesn't work for the admin ID, of course.)